Добавление аватара пользователя

NadirCianid opened the pull request.

NadirCianid committed:
Реализовал UserController

NadirCianid committed:
Реализовал работу с аватарками пользователей

NadirCianid committed:
Реализовал работу с аватарками пользователей

NadirCianid committed:
Реализовал работу с аватарками пользователей

NadirCianid committed:
Rebase merge conflict solved

RickHammerson commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,50 @@
+@RequestMapping("/users")
+@RequiredArgsConstructor
+public class UserController {
+    private static final double MAX_SIZE = 5_242_880L;
```
В имя можно вставить уточнение, что это максимальный размер аватарки или картинки. Можем загружать же ещё что-то и для него будет уже другой макс размер

NadirCianid replied:
в сервисе пользователей только аватарки. Но уточнение добавлю

RickHammerson replied:
Это пока что аватарки. Возможно, будут другие файлы, по типу сертификатов или ещё чего. Лучше указывать явно, к чему это относится

RickHammerson commented on src/main/java/school/faang/user_service/exception/ExceptionMessage.java:
```diff
@@ -11,7 +11,10 @@ public enum ExceptionMessage {
     USER_UNFOLLOWING_HIMSELF_EXCEPTION("The user cannot unfollow himself."),
-    REPEATED_SUBSCRIPTION_EXCEPTION("The user cannot follow another user twice.");
+    REPEATED_SUBSCRIPTION_EXCEPTION("The user cannot follow another user twice."),
+    TOO_LARGE_FILE_EXCEPTION("The uploaded file can have a maximum size of no more than 5 MB."),
```
Лучше выносить в начало то, что с какой сущностью произошла ошибка, а потом причину. Например, у тебя выше есть ошибка
USER_UNFOLLOWING_HIMSELF_EXCEPTION - юзер, сам себя фолловит,понял
TOO_LARGE_FILE_EXCEPTION - а тут придется почитать дальше. Можно FileSizeEception или ещё чего

RickHammerson commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -0,0 +1,50 @@
+    @GetMapping("/{userId}/pic")
+    public ResponseEntity<byte[]> downloadUserPic(@PathVariable Long userId) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.IMAGE_JPEG);
```
Создание хедера можно вынести в приватный метод

NadirCianid replied:
с какой целью? Это единственное место, где он используется и он не слишком много места занимает по-моему

RickHammerson commented on src/main/java/school/faang/user_service/service/S3Service.java:
```diff
@@ -0,0 +1,76 @@
+        return objectMetadata;
+    }
+
+    private AmazonS3 getClient() {
```
Можно вынести в конфигурационный класс и внедрить как зависимость

RickHammerson commented on src/main/java/school/faang/user_service/service/S3Service.java:
```diff
@@ -0,0 +1,76 @@
+@Service
+public class S3Service {
+    private AmazonS3 s3client;
+    @Value("${services.s3.endpoint}")
```
Если интересно, можешь почитать про @ConfigurationProperties аннотацию. Может заменить все эти поля на один объект, который сам будет заполняться переменными окружения с определенным префиксом в названиии

RickHammerson commented on src/main/java/school/faang/user_service/service/user/ImageProcessor.java:
```diff
@@ -0,0 +1,97 @@
+     */
+    public ByteArrayOutputStream getImageOS(Image image) {
+        try (ByteArrayOutputStream bigPicOS = new ByteArrayOutputStream()) {
+            ImageIO.write((RenderedImage) image, "jpeg", bigPicOS);
```
"jpeg" в константу

RickHammerson commented on src/main/java/school/faang/user_service/service/user/ImageProcessor.java:
```diff
@@ -0,0 +1,97 @@
+        BufferedImage bufferedBigPic = getScaledImage(originalPic, bigPicWidth, bigPicHeight);
+        BufferedImage bufferedSmallPic = getScaledImage(originalPic, smallPicWidth, smallPicHeight);
+
+        return List.of(bufferedBigPic, bufferedSmallPic);
```
Думаю, можно сделать отдельный объект для хранения вместо листа

RickHammerson commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,78 @@
+
+        try (InputStream userPicIS = s3Service.downloadFile(user.getUserProfilePic().getFileId())) {
+            return userPicIS.readAllBytes();
+
```
лишний пробел

RickHammerson commented on src/main/java/school/faang/user_service/service/user/ImageProcessor.java:
```diff
@@ -0,0 +1,97 @@
+        int bigPicHeight;
+
+        int smallPicWidth;
+        int smallPicHeight;
```
Заметь, оч много полей, значит можно выделить в отдельный объект, в конструкторе которого уже передаешь BufferedImage originalPic, а он сам обрабатывает для тебя всё и сетит в свои поля, которые ты потом можешь забрать

NadirCianid committed:
исправил комментарии (вынес создание клиента AmazonS3 в отдельный класс)

NadirCianid committed:
исправил комментарии (вынес создание клиента AmazonS3 в отдельный класс)

NadirCianid committed:
Добавил логирование

NadirCianid committed:
Добавил проверку наличия аватара перед удалением и получением, а также добавил удаление предыдущего аватара перед загрузкой нового

RickHammerson commented on src/main/java/school/faang/user_service/service/amazonS3/S3Config.java:
```diff
@@ -0,0 +1,39 @@
+    private String bucketName;
+
+    @PostConstruct
+    public void setUp() {
```
Лучше разделить ответственность. А то это и конфиг, и создание клиента
Можно создать его без private AmazonS3 s3client, а потом создать как @Bean в конфигурационном файле, передавая ему в параметр метода твой S3Config.В теле метода создаешь клиент и возвращаешь его. В итоге у тебя получается уже настроенный бин типа AmazonS3, который ты можешь автовайрить в другие классы
В добавок к этому скажу, что если оставить как у тебя сейчас есть, то тебе нужно будет какждый раз инжектить в свои классы конфиг, в котором ты будешь вызывать ТОЛЬКО получение клиента, а отсальные переменные использовать никогда не будут. Это будет нагромождать код

RickHammerson commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -76,9 +87,17 @@ public byte[] downloadUserPic(Long userId) {
+
+            throw new DataGettingException(USER_AVATAR_ABSENCE_EXCEPTION.getMessage());
+        }
+
```
лишний пробел

Sabina-Vishnya reviewed (commented):
пересечение с задачкой BJS2-6592, а там комменты уже есть, тут они такие же почти, просто здесь кода меньше. 
и кстати по предыдущим пр-ам, тоже комменты пересекаются

учти и просто здесь тоже поправь, либо что-то одно раньше, если поправишь, то в другом пр-е надо будет обновиться и могут вылезти конфлиты. 
Ну а чтобы тебе одно и тоже не делать в ветках разных, можно в одной ветке исправить, затем когда вольешь (в другой ветке, какой-то код пересекается с текущим) поправить с изменениями, например, эту

NadirCianid committed:
Написал тесты

NadirCianid committed:
Написал тесты

NadirCianid committed:
Подготовка к merge

NadirCianid committed:
Merge branch 'phoenix-master-bc4' into feature-BJS2-6537

# Conflicts:
#	src/main/java/school/faang/user_service/controller/user/UserController.java
#	src/main/java/school/faang/user_service/service/user/UserService.java

NadirCianid committed:
merge fix

NadirCianid committed:
Написал тесты

NadirCianid committed:
Merge branch 'feature-BJS2-6774' into feature-BJS2-6592

NadirCianid committed:
Merge pull request #1082 from CorporationX/feature-BJS2-6592

Интеграция с генератором случайных аватаров юзера

NadirCianid committed:
Post-merge refactoring:
- Восстановил работу метода по получению случайного аватара
- Удалил устаревший файл

NadirCianid committed:
Post-merge refactoring:
- Восстановил работу метода по получению случайного аватара
- Удалил устаревший файл

NadirCianid committed:
Merge pull request #1066 from CorporationX/feature-BJS2-6537

Реализовал UserController для системы постов

Sabina-Vishnya commented on src/main/java/school/faang/user_service/service/user/image/AvatarGeneratorService.java:
```diff
@@ -0,0 +1,38 @@
+        log.info("Avatar generated successfully.");
+        return ipImage.getBufferedImage();
+    }
+
```
отступы

Sabina-Vishnya commented on src/main/java/school/faang/user_service/entity/goal/GoalInvitation.java:
```diff
@@ -10,6 +10,7 @@
 import school.faang.user_service.entity.User;
 
 import java.time.LocalDateTime;
+import java.util.Objects;
```
лишний импорт

Sabina-Vishnya commented on src/main/java/school/faang/user_service/exception/ExceptionMessage.java:
```diff
@@ -5,13 +5,20 @@ public enum ExceptionMessage {
     INVALID_EVENT_START_DATE_EXCEPTION("The event start date must be non null and cannot be in the past."),
```
класс много где используется и был наполнен еще новыми значениями, и появился новый класс MessageForGoalInvitationService, теперь не очень понятно почему, тогда где-то разделение есть, а где-то нет
это будет рефачиться?

NadirCianid replied:
Надеюсь. Задачку с GoalInvitations делал Вадим. Я видимо неправильно rebase делал при обновлении ветки

NadirCianid committed:
Merge remote-tracking branch 'origin/feature-BJS2-6774' into feature-BJS2-6774

# Conflicts:
#	src/main/java/school/faang/user_service/controller/goal/GoalInvitationController.java
#	src/main/java/school/faang/user_service/exception/ErrorResponse.java
#	src/test/java/school/faang/user_service/controller/GoalInvitationControllerTest.java
#	src/test/java/school/faang/user_service/service/filter/InvitedIdFilterTest.java
#	src/test/java/school/faang/user_service/service/filter/InvitedNamePatternFilterTest.java
#	src/test/java/school/faang/user_service/service/filter/InviterIdFilterTest.java
#	src/test/java/school/faang/user_service/service/filter/InviterNamePatternFilterTest.java
#	src/test/java/school/faang/user_service/service/filter/RequestStatusFilterTest.java

NadirCianid committed:
Rebase conflicts fix
